### PR TITLE
Feature refactor cleanup analyzer

### PR DESCRIPTION
**Pull request by sopel39**
This PR cleans up and formalizes analyzer and planner in relation to scopes. The main goal is to be able to use hierarchal scopes during planning (currently only top scope is used for field resolution) and to hide (when required) parent scope fields. This is required to cleanly fix: https://github.com/prestodb/presto/issues/7142

Various cleanups/improvments:
- added `OrderBy` AST node
- removed unnecessary/dead code
- hide `StatementAnalyzer` visitor methods (previously external classes were directly calling on them. However, only `StatementAnalyzer` should decide what visitors context is)
- always build entire `WithQueries` map in child scope. This allows to decouple `Scope.parent` from `WithQuery` resolution (those are not the same)
- reintroduce concept of query boundary. Now `ResolvedField.local=false` means resolved field is from outer query scope
- refactor `StatementAnalyzer` so that context scope is always query local to the node, while there is also `outerQueryScope`, e.g: provided by subquery parent
- add support for hierarchical scopes during planning (redefine how fields are resolved in `TranslationMap`, make `QuerySpecification` scope not to link to it's local query parent (hide parent scope fields))

### CODE REVIEW

**Comment by sopel39** [comment]
Applied comments from: https://github.com/prestodb/presto/issues/7227

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 1]
```diff
@@ -191,7 +196,14 @@ public boolean containsNamedQuery(String name)
...
         public Scope build()
         {
-            return new Scope(parent, relationType, namedQueries);
+            Map<String, WithQuery> allNamedQueries = ImmutableMap.<String, WithQuery>builder()
+                    .putAll(namedQueries)
+                    .putAll(parentNamedQueries.map(queries -> queries.entrySet().stream()
+                            .filter(entry -> !namedQueries.containsKey(entry.getKey()))
+                            .collect(toMap(Map.Entry::getKey, Map.Entry::getValue)))
+                            .orElse(emptyMap()))
+                    .build();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: What you would say for something like that:

```
allNamedQueries = Stream.concat(
  namedQueries.stream(),
  parentNamedQueries.map(queries -> queries.entrySet().stream()
           .filter(entry -> !namedQueries.containsKey(entry.getKey()))
      .orElse(emptyMap().stream)
).collect(toImmutableMap(..));
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 2]
```diff
@@ -178,6 +182,21 @@ public Builder withParent(Scope parent)
         {
             checkArgument(!this.parent.isPresent(), "parent is already set");
             this.parent = Optional.of(parent);
+            withParentNamedQueries(parent);
+            return this;
+        }
+
+        public Builder withOuterQueryParent(Scope parent)
+        {
+            checkArgument(!this.parent.isPresent(), "parent is already set");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: call `withParent` here

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 3]
```diff
@@ -1914,24 +1931,49 @@ private void analyzeOrderBy(Query node, Scope scope)
...
+        {
+            return scopeBuilder(parentScope).build();
+        }
+
+        private Scope.Builder scopeBuilder(Optional<Scope> parentScope)
+        {
+            Scope.Builder scopeBuilder = Scope.builder();
+
+            if (parentScope.isPresent()) {
+                scopeBuilder.withParent(parentScope.get());
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: if you have `parentScope` then you simply ignore `outerQueryScope`?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -1914,24 +1931,49 @@ private void analyzeOrderBy(Query node, Scope scope)
...
-                    .build();
+            return createAndAssignScope(node, parentScope, new RelationType(fields));
+        }
+
+        private Scope createAndAssignScope(Node node, Optional<Scope> parentScope, RelationType relationType)
+        {
+            Scope.Builder scopeBuilder = scopeBuilder(parentScope);
+            scopeBuilder.withRelationType(relationType);
+
+            Scope scope = scopeBuilder.build();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Rewrite this to
```
Scope scope = scopeBuilder(parentScope)
  .withRelationType(relationType)
  .build();
```

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 5]
```diff
@@ -207,28 +208,45 @@ public StatementAnalyzer(
...
+
+    private void analyzeWhere(Node node, Scope outerQueryScope, Expression predicate)
+    {
+        Visitor visitor = new Visitor(Optional.of(outerQueryScope));
+        visitor.analyzeWhere(node, outerQueryScope, predicate);
     }
 
     private class Visitor
-            extends DefaultTraversalVisitor<Scope, Scope>
+            extends DefaultTraversalVisitor<Scope, Optional<Scope>>
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I wonder if it can be `DTV<Scope, Void>`

Typically when you process a node then you have scopes of its children so you can calculate your scope and return. What are the cases when you need to calculate parent scope which is passed to its children?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 6]
```diff
@@ -576,16 +576,13 @@ protected Scope visitExplain(Explain node, Optional<Scope> scope)
         protected Scope visitQuery(Query node, Optional<Scope> scope)
         {
             Scope withScope = analyzeWith(node, scope);
-            Scope queryScope = Scope.builder()
-                    .withParent(withScope)
-                    .build();
-            Scope queryBodyScope = process(node.getQueryBody(), queryScope);
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: put into separate commit as it does seem to be related

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -1975,5 +1976,17 @@ else if (outerQueryScope.isPresent()) {
...
         }
+
+        private Scope.Builder scopeBuilderWithParentNamedQueries(Scope namedQueriesParent)
+        {
+            Scope.Builder scopeBuilder = Scope.builder()
+                    .withParentNamedQueries(namedQueriesParent);
+
+            if (outerQueryScope.isPresent()) {
+                scopeBuilder.withOuterQueryParent(outerQueryScope.get());
+            }
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]:
```
outerQueryScope.ifPresent(scopeBuilder::withOuterQueryParent)
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java [hunk 8]
```diff
@@ -260,9 +260,7 @@ private PlanBuilder planFrom(QuerySpecification node)
     private RelationPlan planImplicitTable(QuerySpecification node)
     {
         List<Expression> emptyRow = ImmutableList.of();
-        Scope scope = Scope.builder()
-                .withParent(analysis.getScope(node))
-                .build();
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: as separate commit

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestScope.java [hunk 9]
```diff
@@ -38,7 +38,7 @@ public void test()
 
         Field innerColumn2 = Field.newQualified(QualifiedName.of("inner", "column2"), Optional.of("c2"), BIGINT, false, Optional.empty(), false);
         Field innerColumn3 = Field.newQualified(QualifiedName.of("inner", "column3"), Optional.of("c3"), BIGINT, false, Optional.empty(), false);
-        Scope inner = Scope.builder().withParent(outer).withRelationType(new RelationType(innerColumn2, innerColumn3)).build();
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestScope.java]: can you please add tests for hierarchy?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 10]
```diff
@@ -39,6 +39,7 @@
 public class Scope
 {
     private final Optional<Scope> parent;
+    private final boolean queryBoundary;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: I think it should the implementation of `resolve` method

In case of:
```
- outerQueryScope with [a, b]
  - scope [b]
    - scope [a]
```

When you look for `b` from bottom most scope then I think it should return `b` from the `outerQueryScope`. Unless it is hierarchical scope (introduced later), but in this case all the scope should have queryBoundary always set to `true`, unless they are hierarchical.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 11]
```diff
@@ -208,1717 +207,1731 @@ public StatementAnalyzer(
...
+    private void analyzeWhere(Node node, Scope scope, Expression predicate)
     {
-        QualifiedObjectName targetTable = createQualifiedObjectName(session, insert, insert.getTarget());
-        if (metadata.getView(session, targetTable).isPresent()) {
-            throw new SemanticException(NOT_SUPPORTED, insert, "Inserting into views is not supported");
+        Visitor visitor = new Visitor();
+        visitor.analyzeWhere(node, scope, predicate);
+    }
+
+    private class Visitor
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I don't think it should be static (see `ExpressionAnalyzer.Visitor`). It uses fields from outer class

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 5 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I will add comment as discussed

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java (hunk 10 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: discussed offline

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 12]
```diff
@@ -152,17 +158,14 @@ private static boolean isColumnReference(QualifiedName name, RelationType relati
...
-        }
-
         return Optional.empty();
     }
 
     public static final class Builder
     {
         private RelationType relationType = new RelationType();
         private final Map<String, WithQuery> namedQueries = new HashMap<>();
+        private Optional<Map<String, WithQuery>> parentNamedQueries = Optional.empty();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: Optional allows validation if the parentNamedQueries were explicitly provided.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: `withParent` does copy parent named queries too, while when setting outer scope we wan't to provide different named queries parent

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 3 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: yes, parent scope (local query scope hierarchy) should already have outer scope as it's ancestor. Added comment and `checkState` in analyze method

File: presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java (hunk 8 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/QueryPlanner.java]: without this change tests fail. The scope here shouldn't have any parent.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java (hunk 1 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: I just extracted:
```
            Map<String, WithQuery> parentVisibleNamedQueries = parentNamedQueries.map(
                    queries -> queries.entrySet().stream()
                            .filter(entry -> !namedQueries.containsKey(entry.getKey()))
                            .collect(toMap(Map.Entry::getKey, Map.Entry::getValue)))
                    .orElse(emptyMap());
```

**Comment by sopel39** [comment]
applied comments

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 13]
```diff
@@ -69,6 +69,19 @@ private Scope(
         }
     }
 
+    public Optional<Scope> getOuterParent()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: `getOuterQueryParent`?

**Comment by sopel39** [comment]
Superseded by: https://github.com/prestodb/presto/pull/7315

**Comment by kokosing** [comment]
I would not close this. I think it is much easier to merge and review it partially. It will be easier for you either to rebase it partially.

**Comment by sopel39** [comment]
I am not sure. Isn't it simpler to just keep linear history and address comments in one PR. Anyway, all commits are compilable and all tests should pass. Depends on @martint I guess. What do you think @martint, should I keep two PRs (one for refactor, one for fix: https://github.com/prestodb/presto/pull/7315) with one rebased on top of the other?
